minicriu-client.c

```diff
@@ -102,12 +144,295 @@ static int writefile(const char *file, const char *buf, size_t len) {
+
+		if (cur_nhdr->n_namesz % MC_NOTE_PADDING != 0) {
+			int padding = align_up(bytesWritten, MC_NOTE_PADDING) - bytesWritten;
+			fwrite(paddingData, padding, 1, coreFile);
+			bytesWritten += padding;
+			notes_size += padding;
+		}
+
+		fwrite(&prstatus[i], sizeof(struct elf_prstatus), 1, coreFile);
+		bytesWritten += sizeof(struct elf_prstatus);
+		notes_size += sizeof(struct elf_prstatus);
+
+		if (cur_nhdr->n_descsz % MC_NOTE_PADDING != 0) {
+			int padding = align_up(bytesWritten, MC_NOTE_PADDING) - bytesWritten;
+			fwrite(paddingData, padding, 1, coreFile);
+			bytesWritten += padding;
+			notes_size += padding;
+		}
+	}
+
+	// Write NT_FILE
+	Elf64_Nhdr *cur_nhdr = &nhdr[thread_counter];
+	cur_nhdr->n_namesz = sizeof(owner);
+	cur_nhdr->n_descsz = nt_file.descsz;
+	cur_nhdr->n_type = NT_FILE;
+
+	fwrite(cur_nhdr, sizeof(Elf64_Nhdr), 1, coreFile);
+	bytesWritten += sizeof(Elf64_Nhdr);
+	notes_size += sizeof(Elf64_Nhdr);
+
+	fwrite(owner, sizeof(owner), 1, coreFile);
+	bytesWritten += sizeof(owner);
+	notes_size += sizeof(owner);
+
+	if (cur_nhdr->n_namesz % MC_NOTE_PADDING != 0) {
+		int padding = align_up(bytesWritten, MC_NOTE_PADDING) - bytesWritten;
+		fwrite(paddingData, padding, 1, coreFile);
+		bytesWritten += padding;
+		notes_size += padding;
+	}
+
+	fwrite(&nt_file, sizeof(nt_file.count) + sizeof(nt_file.page_size), 1, coreFile);
+	fwrite(&nt_file.filemaps, sizeof(struct filemap), nt_file.count, coreFile);
+	for (int i = 0; i < nt_file.count; i++) {
+		fputs(nt_file.filepath[i], coreFile);
+		fputc('\0', coreFile);
+	}
+
+	if (cur_nhdr->n_descsz % MC_NOTE_PADDING != 0) {
+		int padding = align_up(bytesWritten, MC_NOTE_PADDING) - bytesWritten;
+		fwrite(paddingData, padding, 1, coreFile);
+		bytesWritten += padding;
+		notes_size += padding;
+	}
+
+	// Write PT_LOAD
+	for (int i = 1; i < phnum; i++) {
+		if (phdr[i].p_filesz != 0) {
+			int padding = phdr[i].p_offset - (phdr[i - 1].p_offset + phdr[i - 1].p_filesz);
+			if (padding > 0) {
+				fwrite(paddingData, sizeof(paddingData), padding / sizeof(paddingData), coreFile);
+				fwrite(paddingData, padding % sizeof(paddingData), 1, coreFile);
+			}
+
+			int written = fwrite((void *)phdr[i].p_vaddr, 1, phdr[i].p_filesz, coreFile);
+
+			if (written != phdr[i].p_filesz) {
+				perror("Failed write map content");
+
+				// As a temporary solution we fill the unwritten data with zeros
+				int leftData = phdr[i].p_filesz - written;
+				int n = leftData / sizeof(paddingData);
+
+				int writtenZeroes = fwrite(paddingData, sizeof(paddingData), n, coreFile);
+				if (writtenZeroes != n) {
+					perror("Failed replace map content with zeroes. Failed to create checkpoint.");
+					fclose(proc_maps);
+					return 1;
+				}
+
+				leftData = leftData % sizeof(paddingData);
+				if (leftData) {
+					if (fwrite(paddingData, leftData, 1, coreFile)) {
+						perror("Failed replace map content with zeroes. Failed to create checkpoint.");
+						fclose(proc_maps);
```

Comment:
I guess you want to close `coreFile` here, right?
Maybe it'd be better to refactor this func into smaller blocks? It helps to avoid such mistakes.

Reply:
It really may be useful to split the function into two parts: one for reading data from `/proc/self/maps` and another to create `minicriu_core`.

I'll get into it this weekend.